### PR TITLE
MVPoly: improve the library + fix sub in monomials

### PR DESCRIPTION
**Pull request by dannywillems**
See individual commits.
It is mostly from https://github.com/o1-labs/proof-systems/pull/2702/ to split it in small target chunks.
Built on top of https://github.com/o1-labs/proof-systems/pull/2710.

### CODE REVIEW

File: mvpoly/src/monomials.rs [hunk 1]
```diff
@@ -427,12 +434,72 @@ impl<const N: usize, const D: usize, F: PrimeField> MVPoly<F, N, D> for Sparse<F
...
+        let exponents: [usize; N] = std::array::from_fn(|i| if i == var_usize { 1 } else { 0 });
+        monomials.insert(exponents, F::one());
+        Self { monomials }
+    }
+
     fn from_expr<Column: Into<usize>, ChallengeTerm: Clone>(
         expr: Expr<ConstantExpr<F, ChallengeTerm>, Column>,
     ) -> Self {
-        prime::Dense::from_expr(expr).into()
+        match expr {
```
Comment (dannywillems) [inline mvpoly/src/monomials.rs:448]: I'm refactoring this for info. Coming in a small future PR. It is the same than prime.rs. We can generalize it. We only need to deal with the variables.

Comment (dannywillems) [reply mvpoly/src/monomials.rs:448]: See https://github.com/o1-labs/proof-systems/pull/2714

File: mvpoly/src/prime.rs [hunk 2]
```diff
@@ -239,15 +242,26 @@ impl<F: PrimeField, const N: usize, const D: usize> MVPoly<F, N, D> for Dense<F,
                     let degree = naive_prime_factors(*idx, &mut prime_gen)
                         .iter()
                         .fold(0, |acc, (_, d)| acc + d);
-                    if degree > max_degree {
+                    if degree > max_degree || rng.gen_range(0..10) == 0 {
+                        // Adding zero coefficients one every 10 monomials
                         F::zero()
                     } else {
                         F::rand(rng)
```
Comment (marcbeunardeau88) [inline mvpoly/src/prime.rs:249]: I'd simply replace this line with a generator biased toward 0

File: mvpoly/tests/monomials.rs [hunk 3]
```diff
@@ -381,6 +381,36 @@ fn test_mvpoly_compute_cross_terms_degree_six() {
     assert_eq!(lhs, rhs);
 }
 
+// The cross-terms of a sum of polynomials is the sum of the cross-terms, per
+// power.
+#[test]
+fn test_mvpoly_pbt_cross_terms_addition() {
```
Comment (marcbeunardeau88) [inline mvpoly/tests/monomials.rs:340]: I'd check p_1 + lambda p_2 to check the full linearity property

File: mvpoly/src/prime.rs (hunk 2 above)
Comment (dannywillems) [reply mvpoly/src/prime.rs:249]: > with a generator biased toward 0

How would you do that?

File: mvpoly/tests/monomials.rs (hunk 3 above)
Comment (dannywillems) [reply mvpoly/tests/monomials.rs:340]: Good idea!

File: mvpoly/src/prime.rs (hunk 2 above)
Comment (marcbeunardeau88) [reply mvpoly/src/prime.rs:249]: generate a nb between 0-10, if that number is 10 return 0, otherwise generate uniformly a field elment

Comment (dannywillems) [reply mvpoly/src/prime.rs:249]: But that's what the code does, doesn't?

File: mvpoly/tests/monomials.rs (hunk 3 above)
Comment (dannywillems) [reply mvpoly/tests/monomials.rs:340]: Well no. It isn't true.

**Pull request merged**